authorization/http.go

```diff
@@ -31,6 +46,49 @@ func WithData(ctx context.Context, data string) context.Context {
 	return context.WithValue(ctx, authorizationDataKey, data)
 }
 
+// GetSelectorsInfo extracts the query namespaces from the provided context.
+func GetSelectorsInfo(ctx context.Context) (*SelectorsInfo, bool) {
+	value := ctx.Value(authorizationSelectorsKey)
+	namespaces, ok := value.(*SelectorsInfo)
+
+	return namespaces, ok
+}
+
+// WithSelectorsInfo extends the provided context with the query namespaces.
+func WithSelectorsInfo(ctx context.Context, info *SelectorsInfo) context.Context {
+	return context.WithValue(ctx, authorizationSelectorsKey, info)
+}
+
+// WithLogsQuerySelectorsExtractor returns a middleware that, when enabled, tries to extract
+// the list of namespaces it queries from a LogQL expression.
+func WithLogsQuerySelectorsExtractor(selectorLabels []string) func(http.Handler) http.Handler {
+	enabled := len(selectorLabels) > 0
+
+	selectorLabelMap := make(map[string]bool, len(selectorLabels))
+	for _, l := range selectorLabels {
+		selectorLabelMap[l] = true
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !enabled {
+				next.ServeHTTP(w, r)
+
+				return
+			}
+
+			selectorsInfo, err := extractQuerySelectors(selectorLabelMap, r.URL.Query())
+			if err != nil {
+				httperr.PrometheusAPIError(w, fmt.Sprintf("error extracting query selectors: %s", err), http.StatusInternalServerError)
+
+				return
+			}
+
+			next.ServeHTTP(w, r.WithContext(WithSelectorsInfo(r.Context(), selectorsInfo)))
```

Comment:
Since this package is signal unaware, I suggest to add another parameter `resource string` to this func (See `authorization.WithAuthorizers`) and use it for a switch case for signals supporting query selector extraction.

Reply:
I don't get this one. Shouldn't it be "signal agnostic" at this point? That's why we removed the `logs` sub-struct from the input-document. In my mind only the "extraction" is signal-dependent (because it needs to know which query language to parse), but the part sending it to OPA can be signal agnostic.